                                     Python Basics:

In [1]:
print("My first example")

My first example


if you want to take notes, just select the markdown from dropdown and the cell now allows you write or take notes and it is not a python code cell.

To execute a cell, press shift enter

Git: To learn more about git , go to git-scm.com & gitimmersion.com


#######################################################################################################################

Naming Conventions: 
1. variable name: all lowercase and underscore between words
2. function names: same as variable names
3. class names follow camelcasing .i.e first letters of each word in capital.

Below is the list of naming conventions:

module_name, package_name, ClassName, method_name, ExceptionName, function_name, GLOBAL_CONSTANT_NAME, global_var_name, instance_var_name, function_parameter_name, local_var_name.

Topic 1: Python Objects and DataTypes

DataTypes: Python has 8 datatypes which you can remember like BLIFTDSS stands for Boolean,Lists,Integer,Float,Tuples,Dictionary,Strings,Set

Variable Assignments: 
Rules:
1. It cannot start with a number
2. It cannot have space in between instead use underscore
3. cant use any special symbols and inbuild keywords already present like 'int', 'True', 'global' etc

Ex: my_dogs = 1

Python uses dynamic typing. Means we can reassign values to the same variable name


In [2]:
number = 10

In [53]:
number = 'ONE'

In [54]:
number

'ONE'

Type function : To find the type of the datatype

In [5]:
type(number)

str

In [6]:
number = [1,2]

In [7]:
number

[1, 2]

In [8]:
type(number)

list

#######################################################################################################################

Strings:
Strings are a seqeunce of characters defined using a single quotes or double quotes.
Strings are ordered sequences and hence we can use indexing and slicing.
Slicing uses the following syntax:
[start:stop:step]

In [9]:
name = 'Prem'

In [10]:
name

'Prem'

In [11]:
my_id = "647508"

In [12]:
my_id

'647508'

In [13]:
type(my_id)

str

In [14]:
name[0]

'P'

In [15]:
name[0:2]

'Pr'

In [16]:
my_id[1::2]

'458'

len() Function: It allows you to find the length of the string


In [17]:
len(my_id)

6

In [18]:
len(name)

4

In [19]:
rev = my_id[::-1]

In [20]:
rev

'805746'

Strings are immutable which means the value of a string cant be changed using indexing[] method.
Hence we have the option of string concatenation. See the examlpe below:

In [1]:
str1 = 'Prem'

In [2]:
str1[3] = 'w'

TypeError: 'str' object does not support item assignment

In the above example , item assigment is not allowed which means string are immutable(values cannot be changed.)
In such case, we use string concatenations using the  '+'  symbol.

In [6]:
str2 = str1[0:3] + 'w'

In [7]:
str2

'Prew'

String Multiplication

In [8]:
letter = 'z'

In [9]:
letter *10

'zzzzzzzzzz'

In [13]:
list1 = [0,1,[2]]

In [14]:
list1*3

[0, 1, [2], 0, 1, [2], 0, 1, [2]]

String concatenation is not possible when you try to do it with a number and a string. In such case you to do type casting.

In [15]:
2 + 3

5

In [16]:
'2'+'3'

'23'

In [17]:
'2'+ 3

TypeError: can only concatenate str (not "int") to str

In [18]:
'2'+str(3)

'23'

If you see in the above example1, it is just a normal addition where 2 +3 =5. Example 2 is concatenation between 2 strings. Hence we get the output as '23' and not 5. In this case it performs string concatenation and not addition.  

In Example 3: we are trying to concatenate a number and a string , which gives an error since concatenation is possible only between the strings. Hence a number cant be concatenated with a string.In such case we need to use type casting.

Type Casting or Conversion:
Converting a non-string value to a string and then performing the desired operation. This is with respect to strings. We can do type casting with any datatypes such as intergers,floats,string,etc..

In [20]:
1+ 45.3

46.3

In [21]:
'5'+ 6

TypeError: can only concatenate str (not "int") to str

In [22]:
int('5')+6

11

In [23]:
1+int(45.3)

46

Example 1: we add integer and a floating point number and python performs a automatic type conversion(implicit typecasting) to float and prints the output as 46.3

Example 2: we are trying to concatenate a str and a int value and hence error is thrown. in such case u need to explicitly do type conversion/casting.Hence we convert '5' to int using int('5')

Example3: Is an example of explicit type conversion. First and third example are same but the difference being when we dont do a explicit type conversion, python does a implicit conversion and hence the output of Ex 1: is a floating point number where as when we do a explicit conversion , we get the desired o/p which is a integer value.

String Build-In Methods:

1. Capitalize:

In [30]:
name = 'prem'

In [31]:
name1 = name.capitalize()

In [32]:
name1

'Prem'

In [33]:
name1[3].capitalize()

'M'

1. Capitalize method used to capitalize the first letter of a given string. In example1: we used capitalize method to capitalize the first letter 'p' to 'P'.

2. we can also capitalize a specific letter in the string using indexing. In Example2: we have capitalized the last letter of the string name from 'm' to 'M' 

#######################################################################################################################

2. Count:

In [34]:
name = 'kumar'

In [35]:
name.count('r')

1

In [36]:
name = 'premkumar'

In [38]:
name.count('r')

2

Count() method is used to count the number of times a spcific value is present in a string. 
In Example 1: name has 1 time letter 'r' and in Example 2: name has 2 times letter 'r'.
Hence count() method is used to find the number of occurences of a specfied value in a given string .

#######################################################################################################################

3. endswith()

In [39]:
name = 'prem'

In [40]:
name.endswith('m')

True

In [41]:
name.endswith('e')

False

In [42]:
name[2].endswith('e')

True

In [43]:
name.endswith('M')

False

endswith() method helps us to find whether the string ends with the specified value. if found it returns True and if not found it returns False value.
In ex 1: name 'prem' ends with 'm' and hence it returns True whereas Ex:2 it returns false since it doesnt endwith 'e'.
Ex 4: Since python is case sensitive , when we check if the string is ending with 'M' , it throws false since it ends with 'm' and not 'M'

#######################################################################################################################

4. find(): find returns the index position of the specified value if it is present in the string else it returns '-1'

In [44]:
name = 'prem'

In [45]:
name.find('p')

0

In [46]:
name.find('k')

-1

In [47]:
name.find('r')

1

In [48]:
name.find('e')

2

In [49]:
name.find('M')

-1

In the above examples , if the specified value is found it returns the index position of the value else it returns -1

#######################################################################################################################

5. index(): This method helps in finding the index position of a specified value. if not present it throws error as substring not found.

In [56]:
name.index('p')

0

In [57]:
name.index('k')

ValueError: substring not found

In [58]:
name.index('re')

1

if you are trying to find the index of a substring , in this example it is 're', it returns the index position of the first occurence of the letter.
Hence here value '1' is returned since 'r' index position is 1.

#######################################################################################################################

6.upper() and lower(): These 2 methods are used to convert all the characters of the string to upper or lower letters. we can also convert specific letters with the help of indexing. See the given examples below:

In [60]:
name.upper()

'PREM'

In [61]:
name.lower()

'prem'

In [62]:
name

'prem'

In [63]:
name[3].upper()

'M'

In [64]:
name


'prem'

In [66]:
name1 = name[2:4].upper()

In [67]:
name1

'EM'

In [68]:
name[0:2] + name1

'prEM'

#######################################################################################################################

7. join(): This method helps in joining a iterable objects with the seperator string provided. Iterable objects can be list,tuple,dictionary ,etc

In [69]:
s = '*'

In [70]:
namelist = ['p','r','e','m'] 

In [71]:
s.join(namelist)

'p*r*e*m'

In [72]:
value = ''

In [73]:
value.join(namelist)

'prem'

If you see the above example , it takes the list as its argument and joins it with the seperator string provided. By default 'space' value is used to join. if a sepcific seperator is provided, then it will use that to join the iterables. 
in Example 1: seperator is '*' and in example 2: it is ' ' (space)

#######################################################################################################################

String Formatting: String formating is nothing but how we print the string or interpolate. This is done using 2 menthods:
1. using .format() method
2. using fstrings

.format() syntax:
'string here {} and also here {}'.format('value1', 'value2')

In [74]:
"Hello my name is {} and i am {} years old".format('Prem','30') 

'Hello my name is Prem and i am 30 years old'

In [75]:
"Hello {1}. welcome to {0}".format('PythonProgramming','Prem')

'Hello Prem. welcome to PythonProgramming'

In [76]:
"Hello {name}. Welcome to {subject} Programming".format(name='prem',subject='Python')

'Hello prem. Welcome to Python Programming'

Float formatting: It allows you to format the float value as per the requirement.
Syntax: .format(value:width.precision f), here value is the value that we try to print and width stands for the spacing or in how much width you want the result value to be printed and precision stands for number of decimal values you want it to get printed.

In [81]:
floatvalue1 = 100.12345678

In [82]:
"The value is {f}".format(f=floatvalue1)

'The value is 100.12345678'

In [83]:
"The value is {f:1.3f}".format(f=floatvalue1)

'The value is 100.123'

In [84]:
"The value is {f:1.6f}".format(f=floatvalue1)

'The value is 100.123457'

in the above example , f is the floatvalue and width=1 and precision=6. If we change the width value , extra space gets added or in how much width we want the results to be printed and precision is the number of decimal points we want.

FSTRINGS: we can achieve the above using fstrings. 

In [85]:
name = 'Prem'

In [88]:
print("Hello {}".format(name))

Hello Prem


In [89]:
print(f"Hello {name}")

Hello Prem


The above example shows how to use fstrings to achieve the same result what we got using .format() method.

In [90]:
floatvalue2 = 122.12341234

In [92]:
print("value is {}".format(floatvalue2))

value is 122.12341234


In [93]:
print(f"value is {floatvalue2}")

value is 122.12341234


#######################################################################################################################

LISTS: List is a seqeunce of ordered items in [] brackets that can hold variety of objects.

In [94]:
list1 = [1,'2',3.0,'prem',[2,3]]

list1

Lists are mutable , which means values can be changed after definition. It also supports slicing and indexing.

In [96]:
list1[0] = 'P'

In [97]:
list1

['P', '2', 3.0, 'prem', [2, 3]]

In [98]:
list1[4][1] = 5

In [99]:
list1

['P', '2', 3.0, 'prem', [2, 5]]

In [100]:
list1[1:]

['2', 3.0, 'prem', [2, 5]]

Lists Methods :


1. append(): This method helps to append or add a new element to the end of a list

In [101]:
list1.append(10)

In [102]:
list1

['P', '2', 3.0, 'prem', [2, 5], 10]

In [103]:
list1[4].append(3)

In [104]:
list1

['P', '2', 3.0, 'prem', [2, 5, 3], 10]

Important Note: Based on the object type present in the list , we will be able to call inbuilt methods based on the datatype. For example , in list1[0] the element we have is a string 'P', hence all the inbuilt string methods are available to this. 

 list1[4] is a nested list [2,5] and hence lists inbuilt methods are available to that element in the list1. Hence we are able to apply .append() method to the nested element in the list. 

2. pop(): This method allows us to remove an element from the end of the list.
We can also remove an specific element by specifying the index position to the pop method.
Ex: list1.pop(1)

In [105]:
list1.pop()

10

In [107]:
list1[4].pop()

3

In [108]:
list1[4]

[2, 5]

In [109]:
list1.pop(1)

'2'

3. sort(): This method is used to sort a list. Note: This method doesnt return a value and it just sorts the given list .so assigning the .sort() to some variable will give return None.

In [2]:
list2 = ['b','a','d','c']

In [3]:
sorted_list = list2.sort()

In [4]:
list2

['a', 'b', 'c', 'd']

In [5]:
sorted_list

Here in the above if we try to print the sorted list , we get None value because it doesnot return anything but it sorts the given list.

4. reverse(): This method is used a reverse a given list

In [117]:
list2.reverse()

In [118]:
list2

['d', 'c', 'b', 'a']

5. insert(): This method helps in inserting value at a specific position. Here we have to specify (indexposition,value)

In [120]:
list2.insert(1,2)

In [128]:
list2

[2, 'c', 'b']

In the above example , we inserted at index position 1 the value 2

6.remove(): This method helps to remove any specified element from the list. The difference between pop and remove is , in pop the last element gets removed if position is not mentioned and the particular element gets poped if position is mentioned. In remove(), the specified element we try to remove gets removed , if the element is not present it will throw an error.

In [129]:
list2.remove('d')

ValueError: list.remove(x): x not in list

In [126]:
list2.remove('a')

In [127]:
list2

[2, 'c', 'b']

#######################################################################################################################

Dictionary: Dictionary is an unordered sequence of items specified in key:value pairs inside {key:value}.
Here we retrieve the value by specifying the key in the indexing and it return the value associated with the key if the key is present and if not present it will throw an error.

In [130]:
dict1 = {1:'ONE',2:'TWO'}

In [131]:
dict1[1]

'ONE'

In [132]:
dict1[3]

KeyError: 3

In [133]:
dict1[3] = 'THREE'

In [134]:
dict1

{1: 'ONE', 2: 'TWO', 3: 'THREE'}

It can also have lists,tuples, dictionaries too...

In [135]:
dict1 = {1: 'ONE', 2:[1,2,3], 3:{'k1':100,'k2':200}}

In [136]:
dict1[2][2]

3

In [138]:
dict1[3]['k2']

200

Dictionary Methods:
 .keys(), .values(),.items()

.keys(): it will give all the keys present in a dictionary as a list.

In [139]:
dict1.keys()

dict_keys([1, 2, 3])

.values(): It will give all the values associated with the keys present in a dictionary as a list.

In [140]:
dict1.values()

dict_values(['ONE', [1, 2, 3], {'k1': 100, 'k2': 200}])

.items() gives the keys value pairs present in a dictionary as a tuple.

In [141]:
dict1.items()

dict_items([(1, 'ONE'), (2, [1, 2, 3]), (3, {'k1': 100, 'k2': 200})])

#######################################################################################################################

TUPLES:

Tuples are similar to a list . it is a ordered sequence of items enclosed with in () and the only difference between lists and tuples are that tuples are immutable (values once assigned cannot be changed).

In [142]:
t = (1,2,3)

In [143]:
t

(1, 2, 3)

In [144]:
t[0] = 5

TypeError: 'tuple' object does not support item assignment

Since it is immutable , item assignment through indexing is not possible. values cant be changed once assigned.

In [145]:
t = (1,2,[1,2,3])

In [146]:
t

(1, 2, [1, 2, 3])

In [147]:
t[2][2] = 'prem'

In [148]:
t

(1, 2, [1, 2, 'prem'])

Note: Tuple values cant be changed through indexing whereas entire tuple can be redeclared with new values. In the above example , t is assigned with new values consisting of a list as a third element.

Important: One main thing to note here, tuple values cannot be changed through indexing but if a mutable object like list,set is present inside a tuple, then those values cant be altered using the indexing as shown in the above example.

In the above Ex 2: we have assigned a new value 'prem' to the list which already had the value 3. Hence in tuples item reassigment is possible for the items present in the tuple if it is a mutable object.

TUPLE METHODS:

In [149]:
t.count(2)

1

In [150]:
t.index(1)

0

count() and index() are the only 2 inbuilt methods available for tuple.
count() helps in identifying the number of times an item is present in the tuple and returns the count of that value.
index() helps in identifying the index position of the given item present in the tuple.

In [151]:
t.index('one')

ValueError: tuple.index(x): x not in tuple

In [152]:
t.count('a')

0

NOTE :From the above 2 examples , index will throw an error if the specified item is not present in the tuple whereas count returns 0 if the item is not present in the tuple

#######################################################################################################################

SETS: Sets are unordered collection of unique items placed inside {}

In [154]:
myset = set()

In [155]:
myset

set()

In [157]:
type(myset)

set

In [159]:
myset.add(1)

In [160]:
myset

{1}

In [161]:
myset.add(2,3)

TypeError: add() takes exactly one argument (2 given)

Sets are created using the set method(). New items are added to the set using the add() method.
Note: Only one item at a item can be added to a set using add() method. If you try adding multiple items at a time as shown in the example , you will get error since add method takes only one argument.

In [162]:
mylist = [1,1,2,2,3,3,4,4,5,5]

In [163]:
set(mylist)

{1, 2, 3, 4, 5}

In the above example , mylist has so many duplicate items but when type casted to set , the set has only unique elements .Also the order of the set will not be the same since they are unordered sequences.

In [7]:
myset1 = set()


In [8]:
myset1 = {1,2,3,4}
myset2 = {1,2,3}

In [9]:
myset1.symmetric_difference(myset2)

{4}

In [10]:
myset1.difference(myset2)

{4}

In [11]:
myset2.difference(myset1)

set()

In [12]:
myset2.symmetric_difference(myset1)

{4}

In [13]:
myset1.intersection(myset2)

{1, 2, 3}

In [14]:
myset2.intersection(myset1)

{1, 2, 3}

Inbuilt Methods to remember in a easy way:
--------------------------------------------------------------------
1. Strings : JulIFECC
2. Lists: Parsir
3. Tuple: IC
4. dictionary: vik
5. sets: Raid

------------------------------------------------------------